lib/train_vehicles/stream.ex

```diff
@@ -0,0 +1,88 @@
+defmodule TrainVehicles.Stream do
+  @moduledoc """
+  Uses Api.Stream to subscribe to the Api and receive events.
+  """
+
+  use GenStage
+  alias Phoenix.PubSub
+  alias TrainVehicles.{Parser, TrainVehicle}
+  require Logger
+
+  @type event_type :: :reset | :add | :update | :remove
+
+  @subway_routes [
+    "Blue",
+    "Green-B",
+    "Green-C",
+    "Green-D",
+    "Green-E",
+    "Orange",
+    "Red",
+    "Mattapan"
+  ]
+
+  def start_link(opts) do
+    {name, opts} = Keyword.pop(opts, :name, __MODULE__)
+
+    GenStage.start_link(
+      __MODULE__,
+      opts,
+      name: name
+    )
+  end
+
+  def init(opts) do
+    producer_consumer = Keyword.fetch!(opts, :subscribe_to)
+    broadcast_fn = Keyword.get(opts, :broadcast_fn, &PubSub.broadcast/3)
+    {:consumer, %{broadcast_fn: broadcast_fn}, subscribe_to: [producer_consumer]}
+  end
+
+  def handle_events(events, _from, state) do
+    :ok = Enum.each(events, &send_event(&1, state.broadcast_fn))
+    {:noreply, [], state}
+  end
+
+  defp send_event(
+         %Api.Stream.Event{
+           event: :remove,
+           data: %JsonApi{data: data}
+         },
+         broadcast_fn
+       ) do
+    data
+    |> Enum.map(& &1.id)
+    |> broadcast(:remove, broadcast_fn)
+  end
+
+  defp send_event(
+         %Api.Stream.Event{
+           event: type,
+           data: %JsonApi{data: data}
+         },
+         broadcast_fn
+       ) do
+    data
+    |> Enum.map(&Parser.parse/1)
+    |> Enum.filter(&Enum.member?(@subway_routes, &1.route_id))
+    |> broadcast(type, broadcast_fn)
+  end
+
+  @typep broadcast_fn :: (atom, String.t(), any -> :ok | {:error, any})
+  @spec broadcast([TrainVehicle.t() | String.t()], event_type, broadcast_fn) :: :ok
+  defp broadcast([], _type, _broadcast_fn), do: :ok
+
+  defp broadcast(data, type, broadcast_fn) do
+    TrainVehicles.PubSub
+    |> broadcast_fn.("train_vehicles", {type, data})
```

Comment:
This "train_vehicles` argument is always the same anywhere. Since we control the `broadcast_fn` type, could we just remove that argument from the function? (I know you'd still need to specify it somewhere for the pubsub topic, but it could be specified in the function definition instead of here.)

Reply:
I'm not sure I'm following your suggestion. The `broadcast_fn` is `PubSub.broadcast/3` which takes the topic.

Reply:
Ah, I missed that you were literally using `&Pubsub.broadcast/3`. I thought you were defining your own function somewhere. Nevermind.